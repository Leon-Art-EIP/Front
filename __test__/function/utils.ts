import type { ReactTestRendererJSON } from "react-test-renderer";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";
import { expect } from "vitest";

function toJson(component: renderer.ReactTestRenderer): ReactTestRendererJSON {
  const result = component.toJSON();
  expect(result).toBeDefined();

  return result as ReactTestRendererJSON;
}

export function testSnapshot(element: JSX.Element): void {
  const component = renderer.create(element);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
}
