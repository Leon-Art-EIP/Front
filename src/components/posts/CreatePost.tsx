export default function CreatePost(): JSX.Element {
  return <div className="flex flex-col gap-2"></div>;
}
