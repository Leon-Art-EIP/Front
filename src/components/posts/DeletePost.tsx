import { useState } from "react";
import { Button, Modal } from "../lib";

interface IDeletePostProps {
  onDeletePost: (postId: string) => void;
  postId: string;
}

export default function DeletePost(props: IDeletePostProps): JSX.Element {
  const [isModalOpen, setIsModalOpen] = useState(false);

  const openModal = () => {
    setIsModalOpen(true);
  };

  const closeModal = () => {
    setIsModalOpen(false);
  };

  const deletePost = () => {
    props.onDeletePost(props.postId);
  };

  return (
    <>
      <Modal handleClose={closeModal} isOpen={isModalOpen}>
        <div className="flex flex-col gap-4">
          <h2 className="text-2xl">Supprimer le post</h2>
          <p>Êtes-vous sûr de vouloir supprimer ce post ?</p>
          <div className="flex gap-2 justify-center">
            <Button color="danger" type="button" onClick={deletePost}>
              Supprimer
            </Button>
            <Button color="secondary" type="button" onClick={closeModal}>
              Annuler
            </Button>
          </div>
        </div>
      </Modal>
      <button className="text-sm text-red-400" onClick={openModal}>
        Supprimer
      </button>
    </>
  );
}
