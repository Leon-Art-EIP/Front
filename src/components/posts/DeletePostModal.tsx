import { Button, Modal } from "../lib";

interface IDeletePostModalProps {
  handleClose: () => void;
  handleDelete: () => void;
}

export default function DeletePostModal(props: IDeletePostModalProps): JSX.Element {
  const onCancel = () => {
    props.handleClose();
  };

  const onDelete = () => {
    props.handleDelete();
  };

  return (
    <Modal isOpen handleClose={props.handleClose}>
      <div className="flex flex-col gap-4 p-4 rounded-2xl bg-background max-w-3xl">
        <h2>Voulez-vous vraiment supprimer ce post ?</h2>
        <div className="flex gap-8 justify-center">
          <Button type="button" color="secondary" onClick={onCancel}>
            Annuler
          </Button>
          <Button type="button" color="primary" onClick={onDelete}>
            Oui
          </Button>
        </div>
      </div>
    </Modal>
  );
}
