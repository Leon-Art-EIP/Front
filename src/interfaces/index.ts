export interface ITab {
  name: string;
  href: string;
  loggedIn: boolean;
}

export type ISuccess = {
  token: string;
};

export type IError = {
  errors: {
    type: string;
    msg: string;
    path: string;
    location: string;
    value: string;
  }[];
};
