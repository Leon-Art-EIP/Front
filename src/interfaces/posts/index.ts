export interface INewPost {
  msg: string;
  post: {
    userId: string;
    text: string;
    artPublicationId: string;
    likes: string[];
    _id: string;
    createdAt: string;
    __v: number;
  };
}

export interface IPost {
  _id: string;
  userId: {
    _id: string;
    username: string;
    profilePicture: string;
  };
  text: string;
  artPublicationId?: {
    _id: string;
    name: string;
  };
  likes: string[];
  createdAt: string;
  __v: number;
}

export interface ILikePost {
  msg: string;
  likeStatus: {
    postId: string;
    isLiked: boolean;
    totalLikes: number;
  };
}
