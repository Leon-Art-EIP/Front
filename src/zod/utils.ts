export const getValueOrUndefined = (arg: unknown) =>
  arg === "" || arg === undefined || arg === null ? undefined : arg;
