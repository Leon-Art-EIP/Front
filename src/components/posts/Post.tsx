"use client";

import { Dispatch, SetStateAction, useEffect, useState } from "react";
import { IPost } from "../../interfaces/posts";
import { imageApi } from "../../tools/variables";
import { stringToFrenchDate } from "../../tools/date";
import { IArtPublication } from "../../interfaces/artPublication/artPublication";
import { myFetch } from "../../tools/myFetch";
import { CircularProgress } from "@mui/material";
import PostArtPublication from "./PostArtPublication";
import { DeleteOutline, Favorite, FavoriteBorder } from "@mui/icons-material";
import DeletePostModal from "./DeletePostModal";
import Link from "next/link";

interface IPostProps {
  post: IPost;
  connectedUserId: string;
  onLike: (postId: string) => void;
  onDelete: (postId: string) => void;
  isDeleteLoading: boolean;
}

async function fetchArtPublicationPicture(
  artPublicationId: string,
  setArtPublication: Dispatch<SetStateAction<IArtPublication | null | undefined>>
) {
  const response = await myFetch({
    route: `/api/art-publication/${artPublicationId}`,
    method: "GET",
  });

  if (response.ok) {
    const data = response.json as IArtPublication;

    setArtPublication(data);
  } else {
    setArtPublication(undefined);
  }
}

export default function Post(props: IPostProps): JSX.Element {
  const [artPublication, setArtPublication] = useState<IArtPublication | null | undefined>(null);
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);

  const isLiked = props.post.likes.includes(props.connectedUserId);

  useEffect(() => {
    if (props.post.artPublicationId) {
      fetchArtPublicationPicture(props.post.artPublicationId, setArtPublication);
    }
  }, [props.post]);

  const onLike = () => {
    props.onLike(props.post.id);
  };

  const openDeleteModal = () => {
    setIsDeleteModalOpen(true);
  };

  const closeDeleteModal = () => {
    setIsDeleteModalOpen(false);
  };

  const handleDelete = () => {
    props.onDelete(props.post.id);
    setIsDeleteModalOpen(false);
  };

  return (
    <>
      {isDeleteModalOpen && <DeletePostModal handleClose={closeDeleteModal} handleDelete={handleDelete} />}
      <div className="flex flex-col gap-4 bg-secondary p-4 rounded-2xl w-1/2">
        <div className="flex gap-4 items-center justify-between">
          <Link href={`/profile/${props.post.userId}`} className="flex gap-4 items-center">
            <img alt="author" src={`${imageApi}/${props.post.user.profilePicture}`} className="w-8 h-8 rounded-full" />
            <p className="font-semibold flex">{props.post.user.username}</p>
          </Link>
          <p className="text-xs font-light">{stringToFrenchDate(props.post.createdAt)}</p>
        </div>
        <p className="font-light">{props.post.text}</p>
        {props.post.artPublicationId &&
          (artPublication === null ? (
            <CircularProgress size={20} thickness={4} color="primary" />
          ) : (
            <PostArtPublication
              artPublication={{
                _id: props.post.artPublicationId,
                image: artPublication?.image,
              }}
            />
          ))}
        <div className="flex justify-between">
          <button type="button" className="flex items-center gap-1.5" onClick={onLike}>
            {isLiked ? <Favorite /> : <FavoriteBorder />} {props.post.likes.length}
          </button>
          {props.connectedUserId === props.post.userId && (
            <button type="button" onClick={openDeleteModal} disabled={props.isDeleteLoading}>
              {props.isDeleteLoading ? <CircularProgress size={20} thickness={4} /> : <DeleteOutline />}
            </button>
          )}
        </div>
      </div>
    </>
  );
}
