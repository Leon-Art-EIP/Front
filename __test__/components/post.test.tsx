import { test } from "vitest";
import { testSnapshot } from "../function/utils";
import Post from "../../src/components/posts/Post";
import { IPost } from "../../src/interfaces/posts";

const post: IPost = {
  userId: "1",
  text: "text",
  artPublicationId: "1",
  createdAt: "2021-09-01",
  likes: [],
  id: "1",
  user: {
    username: "username",
    profilePicture: "profilePicture",
  },
  artPublication: {
    name: "name",
  },
};

test("Post deleteLoading snapshot", () => {
  testSnapshot(<Post connectedUserId="1" isDeleteLoading onDelete={() => {}} onLike={() => {}} post={post} />);
});

test("Post snapshot", () => {
  testSnapshot(<Post connectedUserId="1" isDeleteLoading={false} onDelete={() => {}} onLike={() => {}} post={post} />);
});
