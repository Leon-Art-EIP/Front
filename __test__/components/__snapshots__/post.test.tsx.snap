// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Post deleteLoading snapshot 1`] = `
<div
  className="flex flex-col gap-4 bg-secondary p-4 rounded-2xl w-1/2"
>
  <div
    className="flex gap-4 items-center justify-between"
  >
    <a
      className="flex gap-4 items-center"
      href="/profile/1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      <img
        alt="author"
        className="w-8 h-8 rounded-full"
        src="undefined/api/profilePicture"
      />
      <p
        className="font-semibold flex"
      >
        username
      </p>
    </a>
    <p
      className="text-xs font-light"
    >
      01/09/2021 à 02:00:00
    </p>
  </div>
  <p
    className="font-light"
  >
    text
  </p>
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
    role="progressbar"
    style={
      {
        "height": 20,
        "width": 20,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20}
        strokeWidth={4}
        style={{}}
      />
    </svg>
  </span>
  <div
    className="flex justify-between"
  >
    <button
      className="flex items-center gap-1.5"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="FavoriteBorderIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
        />
      </svg>
       
      0
    </button>
    <button
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20}
            strokeWidth={4}
            style={{}}
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`Post snapshot 1`] = `
<div
  className="flex flex-col gap-4 bg-secondary p-4 rounded-2xl w-1/2"
>
  <div
    className="flex gap-4 items-center justify-between"
  >
    <a
      className="flex gap-4 items-center"
      href="/profile/1"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      <img
        alt="author"
        className="w-8 h-8 rounded-full"
        src="undefined/api/profilePicture"
      />
      <p
        className="font-semibold flex"
      >
        username
      </p>
    </a>
    <p
      className="text-xs font-light"
    >
      01/09/2021 à 02:00:00
    </p>
  </div>
  <p
    className="font-light"
  >
    text
  </p>
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
    role="progressbar"
    style={
      {
        "height": 20,
        "width": 20,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20}
        strokeWidth={4}
        style={{}}
      />
    </svg>
  </span>
  <div
    className="flex justify-between"
  >
    <button
      className="flex items-center gap-1.5"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="FavoriteBorderIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
        />
      </svg>
       
      0
    </button>
    <button
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
  </div>
</div>
`;
